include/OpenSpaceToolkit/Astrodynamics/EventCondition.hpp

```diff
@@ -0,0 +1,122 @@
+/// state vectors and times
+
+class EventCondition
+{
+   public:
+    enum class Criteria
+    {
+        PositiveCrossing,
+        NegativeCrossing,
+        AnyCrossing,
+        PositiveOnly,
+        NegativeOnly,
+        Undefined
+    };
+
+    /// @brief              Constructor
+    ///
+    /// @code
+    ///                     EventCondition eventCondition = {aName, aCriteria};
+    /// @endcode
+    ///
+    /// @param              [in] aName A string representing the name of the Event Condition
+    /// @param              [in] aCriteria An enum indicating the criteria used to determine the Event Condition
+
+    EventCondition(const String& aName, const Criteria& aCriteria);
+
+    /// @brief              Virtual destructor
+
+    virtual ~EventCondition();
+
+    /// @brief              Output stream operator
+    ///
+    /// @code
+    ///                     std::cout << EventCondition(...) ;
+    /// @endcode
+    ///
+    /// @param              [in] anOutputStream An output stream
+    /// @param              [in] anEventCondition An EventCondition
+    /// @return             A reference to output stream
+
+    friend std::ostream& operator<<(std::ostream& anOutputStream, const EventCondition& anEventCondition);
+
+    /// @brief              Get name of the Event Condition
+    ///
+    /// @return             String representing the name of the Event Condition
+
+    String getName() const;
+
+    /// @brief              Get the criteria of the Event Condition
+    ///
+    /// @return             Enum representing the criteria of the Event Condition
+
+    Criteria getCriteria() const;
+
+    /// @brief              Print the Event Condition
+    ///
+    /// @param              [in, out] anOutputStream The output stream where the Event Condition will be printed
+    /// @param              [in] displayDecorator A boolean indicating whether or not to display decorator during
+    /// printing
+
+    virtual void print(std::ostream& anOutputStream, bool displayDecorator = true) const;
+
+    /// @brief              Check if the Event Condition is satisfied based on current value and previous value
+    ///
+    /// @param              [in] currentValue The current value
+    /// @param              [in] previousValue The previous value
+    ///
+    /// @return             Boolean value indicating if the Event Condition is met
+
+    virtual bool isSatisfied(const Real& currentValue, const Real& previousValue) const;
+
+    /// @brief              Evaluate the Event Condition
+    ///
+    /// @param              [in] aStateVector The current state vector
+    /// @param              [in] aTime The current time
+    ///
+    /// @return             Real number representing the evaluation result of the Event Condition
+
+    virtual Real evaluate(const VectorXd& aStateVector, const Real& aTime) const = 0;
```

Comment:
Yeah that's true. Every condition is defined by this mapping:
StateVector + Time -> a Real value which resolves to the targeted value

Reply:
I know we've already talked about this @vishwa2710 so sorry to bring it back but the more I think we are boxing ourselves.

I really think this should be `evaluate (const State& aState)`

I know the first step is just to numerically propagate to this conditions but there are a bunch of problems that will arise from using `VectorXd` and `Real` as a time offset.

- Time based event conditions (which are many, almost all of them) will now need to carry over a private property storing initial time so that they can convert the offset to an instant.
- No frame-related information. One gets a `VectorXd` but no reference frame attached to this. So eventually this gets either passed or it needs to be 'assumed` by conditions (i.e. hardcoded)

Those are just details, but it overall defeats the purpose of having a `State` concept that already does all this for us. You can pass it a coordinate subset and it will give it to you, in the desired frame (since it has all the info inside).

I think the `EventCondition` interface is great, but this particular signature boxes us into really low level, when we don't really need to.

In the numerical propagator, just create a `State` off of your `VectorXd`, instant and integration reference frame, and pass it along to the EventCondition. Maybe just a tiny bit slowe but for sure a much better dev experience imo.

Reply:
Okay I discussed this with @phc1990 offline and we agree that for now we should move forward as is. Once we land his State + CoordinateBroker MR we can figure out what's the best way to leverage this in the conditions.

The conditions will always need a VectorXd + Double/Real interface as the NumericalSolver is aimed to be agnostic of the Astro stuff once it moves to Math.

In the future if we want, we can subclass the solver in astro and make it state aware if required.

Reply:
> Once we land his State + CoordinateBroker MR we can figure out what's the best way to leverage this in the conditions.

That seems reasonable!

Reply:
> The conditions will always need a VectorXd + Double/Real interface as the NumericalSolver is aimed to be agnostic of the Astro stuff once it moves to Math.

👍